pkg/cmd/bazci/bazci.go

```diff
@@ -268,3 +285,119 @@ func doCopy(src, dst string) error {
+		// ever asked to post, so enforce that on all runs.
+		// The way this env var is made available here is quite tricky. The build
+		// calling this method is usually a build that is invoked from PRs, so it
+		// can't have secrets available to it (for the PR could modify
+		// build/teamcity-* to leak the secret). Instead, we provide the secrets
+		// to a higher-level job (Publish Bleeding Edge) and use TeamCity magic to
+		// pass that env var through when it's there. This means we won't have the
+		// env var on PR builds, but we'll have it for builds that are triggered
+		// from the release branches.
+		if os.Getenv("GITHUB_API_TOKEN") == "" {
+			return errors.New("GITHUB_API_TOKEN must be set")
+		}
+		githubpost.Post(githubPostFormatterName, bytes.NewReader(goTestJsonOutputFileBuf))
+	}
+	return nil
+}
+
+func createTarball() error {
+	buf, err := os.ReadFile(goTestJsonOutputFilePath)
+	if err != nil {
+		return err
+	}
+	f, err := os.Create(filepath.Join(artifactsDir, "test_output.txt"))
+	if err != nil {
+		return err
+	}
+	if _, err := testfilter.FilterAndWrite(bytes.NewReader(buf), f, []string{"convert"}); err != nil {
+		return err
+	}
+
+	tarArgs := []string{
+		"--strip-components=1",
+		"-czf",
+		"full_output.tgz",
+		"full_output.txt",
+		filepath.Base(goTestJsonOutputFilePath),
+	}
+	createTarballCmd := exec.Command("tar", tarArgs...)
+	createTarballCmd.Dir = artifactsDir
+	var errBuf bytes.Buffer
+	createTarballCmd.Stderr = &errBuf
+	fmt.Println("running tar w/ args: ", shellescape.QuoteCommand(tarArgs))
+	if err := createTarballCmd.Run(); err != nil {
+		return errors.Wrapf(err, "StdErr: %s", errBuf.String())
+	}
+	if err := os.Remove(filepath.Join(artifactsDir, "full_output.txt")); err != nil {
+		fmt.Printf("Failed to remove full_output.txt - %v\n", err)
+	}
+	return nil
+}
+
+// Some unit tests test automatic ballast creation. These ballasts can be
+// larger than the maximum artifact size. Remove any artifacts with the
+// EMERGENCY_BALLAST filename.
+func removeEmergencyBallasts() {
+	findCmdArgs := []string{
+		"-name",
+		artifactsDir,
+		"EMERGENCY_BALLAST",
+		"-delete",
+	}
+	findCmd := exec.Command("find", findCmdArgs...)
+	var errBuf bytes.Buffer
+	findCmd.Stderr = &errBuf
+	if err := findCmd.Run(); err != nil {
+		fmt.Println("running find w/ args: ", shellescape.QuoteCommand(findCmdArgs))
+		fmt.Printf("Failed with err %v\nStdErr: %v", err, errBuf.String())
+	}
+}
+
+func processTestJsonIfNeeded(shouldCreateTarball bool) error {
+	if goTestJsonOutputFilePath == "" {
+		return nil
+	}
+	buf, err := os.ReadFile(goTestJsonOutputFilePath)
+	if err != nil {
+		return err
+	}
+	if err := processFailures(buf, filepath.Join(artifactsDir, "failures.txt")); err != nil {
+		return err
+	}
+	if err := postReleaseOnlyFailures(buf); err != nil {
+		return err
+	}
+	if shouldCreateTarball {
+		if err := createTarball(); err != nil {
+			return err
+		}
+	}
+	removeEmergencyBallasts()
```

Comment:
I agree executing this as an unconditional cleanup step regardless of whether the previous steps succeeded or failed would be an improvement.